src/VisualStudio/Core/Def/Snippets/SemanticSnippetExpander.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncCompletion;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Snippets;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+
+namespace Microsoft.VisualStudio.LanguageServices.Snippets;
+
+[Export(typeof(ISemanticSnippetExpander))]
+[Shared]
+internal sealed class SemanticSnippetExpander : ISemanticSnippetExpander
+{
+    private readonly RoslynLSPSnippetParser _snippetParser;
+    private readonly RoslynLSPSnippetConverter _snippetConverter;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public SemanticSnippetExpander(
+        RoslynLSPSnippetParser snippetParser,
+        RoslynLSPSnippetConverter snippetConverter)
```

Comment:
📝 All of these new types are ported from the internal implementation. We can't use the internal implementation because we need the expansion to use our own AbstractSnippetExpansionClient (along with its new formatting support)

Reply:
Can you update the PR description to give a bit more detail about why we had to do this? Or why this isn't just something we're fixing somewhere else?